core-bundle/src/Twig/Interop/ContextFactory.php

```diff
@@ -0,0 +1,159 @@
+<?php
+
+declare(strict_types=1);
+
+/*
+ * This file is part of Contao.
+ *
+ * (c) Leo Feyer
+ *
+ * @license LGPL-3.0-or-later
+ */
+
+namespace Contao\CoreBundle\Twig\Interop;
+
+use Contao\Template;
+
+/**
+ * @experimental
+ */
+final class ContextFactory
+{
+    /**
+     * Create a Twig template context from a @see Template object.
+     */
+    public function fromContaoTemplate(Template $template): array
+    {
+        $context = $template->getData();
+
+        array_walk_recursive(
+            $context,
+            function (&$value, $key): void {
+                if ($value instanceof \Closure) {
+                    $value = $this->getCallableWrapper($value, (string) $key);
+                }
+            }
+        );
+
+        if (!isset($context['Template'])) {
+            $context['Template'] = $template;
+        }
+
+        return $context;
+    }
+
+    /**
+     * Create a Twig template context from an arbitrary object. This will also
+     * make private or protected methods/properties/constants accessible.
+     */
+    public function fromClass(object $object): array
+    {
+        $class = new \ReflectionClass($object);
+
+        $context = array_merge(
+            $class->getConstants(),
+            $class->getStaticProperties(),
+            iterator_to_array($this->getAllMembers($object)),
+        );
+
+        foreach ($class->getMethods() as $method) {
+            $name = $method->getName();
+
+            if (0 === strpos($name, '__')) {
+                continue;
+            }
+
+            $context[$name] = $this->getCallableWrapper($method->getClosure($object), $name);
+        }
+
+        if (!isset($context['data'])) {
+            $context['data'] = $object;
+        }
+
+        return $context;
+    }
+
+    /**
+     * Find all members including those that were dynamically set ($this->foo = 'bar').
+     */
+    private function getAllMembers(object $object): \Generator
+    {
+        // See https://externals.io/message/105697#105697
+        $mangledObjectVars = \function_exists('get_mangled_object_vars') ?
+            get_mangled_object_vars($object) : (array) $object;
```

Comment:
```suggestion
        // See https://externals.io/message/105697#105697
        // Backwards compatibility with PHP < 7.4
        $mangledObjectVars = \function_exists('get_mangled_object_vars') ?
            get_mangled_object_vars($object) : (array) $object;
```
Comment so that we don’t forget to remove this in the future?

Reply:
Sounds good!